src/lib.rs

```diff
@@ -0,0 +1,337 @@
+                    };
+                }
+                _ => return Error::ContentWrong.into_enum(),
+            }
+        }
+        self.into_enum()
+    }
+}
+
+#[derive(Debug)]
+struct BulkBodyParser {
+    expect: usize,
+}
+
+impl BulkBodyParser {
+    fn parse(mut self, buffer: &mut Buffer) -> ParseOutput {
+        let scanned = buffer.consume(self.expect);
+        self.expect -= scanned;
+        if self.expect > 0 {
+            return self.into_enum();
+        }
+
+        let mut data = buffer.pop_scanned_buffer();
+        if data.ends_with(b"\r\n") {
+            data.pop();
+            data.pop();
+            Message::Bulk(Some(data)).into_enum()
+        } else {
+            Error::ContentWrong.into_enum()
+        }
+    }
+}
+
+#[derive(Debug, Default)]
+struct ArraySizeParser(usize);
+
+impl ArraySizeParser {
+    fn parse(mut self, buffer: &mut Buffer) -> ParseOutput {
+        for i in buffer.iter() {
+            match i {
+                x @ b'0'..=b'9' => self.0 = self.0 * 10 + usize::from(x - b'0'),
+                b'\r' => {}
+                b'\n' => {
+                    buffer.pop_scanned_buffer();
+                    return match self.0 {
+                        0 => Message::Array(vec![]).into_enum(),
+                        x => ArrayBodyParser {
+                            message: Vec::with_capacity(x),
+                            expect: x,
+                            current: Box::new(InitParser.into_enum()),
+                        }
+                        .into_enum(),
+                    };
+                }
+                _ => return Error::ContentWrong.into_enum(),
+            }
+        }
+        self.into_enum()
+    }
+}
+
+#[derive(Debug)]
+struct ArrayBodyParser {
+    expect: usize,
+    message: Vec<Message>,
+    current: Box<Parser>,
+}
+
+impl ArrayBodyParser {
+    fn parse(mut self, buffer: &mut Buffer) -> ParseOutput {
+        if self.expect == self.message.len() {
+            return Message::Array(self.message).into_enum();
+        }
+
+        match self.current.parse(buffer).downcast() {
+            ParseOutputType::Err(e) => e.into_enum(),
+            ParseOutputType::Message(m) => {
+                self.message.push(m);
+
+                self.current = Box::new(InitParser.into_enum());
+
+                self.parse(buffer)
+            }
+            ParseOutputType::Parser(p) => {
+                self.current = Box::new(p);
+                self.into_enum()
+            }
+        }
+    }
+}
+
+#[derive(Debug, Default)]
+struct InlineParser;
+
+impl InlineParser {
+    fn parse(self, buffer: &mut Buffer) -> ParseOutput {
+        if buffer.iter().position(|x| x == b'\n') == None {
+            return self.into_enum();
+        }
+
+        let mut data = buffer.pop_scanned_buffer();
+        if data.pop() != Some(b'\n') {
+            return Error::ContentWrong.into_enum();
+        }
+        if data.pop() != Some(b'\r') {
```

Comment:
inline command always ends with `\n` but not usually ends with `\r\n`. That's not error.